app/models/manageiq/providers/embedded_terraform/automation_manager/openstack_credential.rb

```diff
@@ -0,0 +1,131 @@
+      :name       => 'project',
+      :id         => 'project',
+      :maxLength  => 100,
+      :isRequired => true,
+      :validate   => [{:type => 'required'}],
+    },
+    {
+      :component  => 'text-field',
+      :label      => N_('Domain Name'),
+      :helperText => N_('OpenStack domains define administrative boundaries. It is only needed for Keystone v3 authentication URLs'),
+      :name       => 'domain',
+      :id         => 'domain',
+      :maxLength  => 100,
+    },
+    {
+      :component  => 'text-field',
+      :label      => N_('Region Name'),
+      :helperText => N_('The region to create resources on'),
+      :name       => 'region',
+      :id         => 'region',
+      :maxLength  => 100,
+    },
+    {
+      :component      => 'password-field',
+      :label          => N_('CA certificate'),
+      :helperText     => N_('The custom CA certificate when communicating over SSL'),
+      :componentClass => 'textarea',
+      :name           => 'cacert',
+      :id             => 'cacert',
+      :type           => 'password',
+    },
+    {
+      :component      => 'password-field',
+      :label          => N_('Client Certificate'),
+      :helperText     => N_('The client certificate for SSL client authentication.'),
+      :componentClass => 'textarea',
+      :name           => 'clientcert',
+      :id             => 'clientcert',
+      :type           => 'password',
+    },
+    {
+      :component  => 'checkbox',
+      :label      => N_('Trust self-signed SSL certificates ?'),
+      :helperText => N_('Trust self-signed SSL certificates ?'),
+      :name       => 'insecure',
+      :id         => 'insecure',
+    },
+  ].freeze
+
+  API_ATTRIBUTES = (COMMON_ATTRIBUTES + EXTRA_ATTRIBUTES).freeze
+
+  API_OPTIONS = {
+    :type       => 'cloud',
+    :label      => N_('OpenStack'),
+    :attributes => API_ATTRIBUTES
+  }.freeze
+
+  def self.display_name(number = 1)
+    n_('Credential (OpenStack)', 'Credentials (OpenStack)', number)
+  end
+
+  def self.params_to_attributes(params)
+    attrs = super.dup
+
+    if %i[host domain project region cacert clientcert clientkey insecure].any? { |opt| attrs.has_key?(opt) }
+      attrs[:options]         ||= {}
+      attrs[:options][:host]    = attrs.delete(:host)    if attrs.key?(:host)
+      attrs[:options][:domain]  = attrs.delete(:domain)  if attrs.key?(:domain)
+      attrs[:options][:project] = attrs.delete(:project) if attrs.key?(:project)
+      attrs[:options][:region]  = attrs.delete(:region)  if attrs.key?(:region)
+
+      attrs[:options][:cacert]     = attrs.delete(:cacert)     if attrs.key?(:cacert)
+      attrs[:options][:clientcert] = attrs.delete(:clientcert) if attrs.key?(:clientcert)
+      attrs[:options][:clientkey]  = attrs.delete(:clientkey)  if attrs.key?(:clientkey)
+
+      attrs[:options][:insecure] = attrs.delete(:insecure) if attrs.key?(:insecure)
+    end
+
+    attrs
```

Comment:
I think something like this does what you're trying to do:

```suggestion
    if attrs.keys & %i[host domain project region cacert clientcert clientkey insecure]
      attrs.slice!(:host, :domain, :project, :region, :cacert, :clientkey, :insecure)
    end

    {:options => attrs}
```

Reply:
Are there other values you're concerned about retaining?  The slice option will only remove all key/values that aren't in the list so if you have others that aren't in that list, they'll be lost.

Reply:
I think `attrs` will have more than just these keys so we can't replace the top level hash with just `{:options =>`
What about something like

```
options = super.dup
attrs = options.slice!(:host, :domain, :project, :region, :cacert, :clientkey, :insecure)
attrs.merge!(:options => options) unless options.blank?
attrs
```

`attrs` will now have the rest of the keys that aren't in the `slice!` which should be the rest of the top level keys.

Reply:
Note, we have this pattern in `params_to_attributes` to pull out specific keys from `attrs` and store them in the `:options` subkey instead.  Perhaps extract a method `move_keys_to_options` from above.

Reply:
Or not... this one is the most complicated one.

Reply:
> Note, we have this pattern in params_to_attributes to pull out specific keys from attrs and store them in the :options subkey instead. Perhaps extract a method move_keys_to_options from above.

Yeah I wonder if maybe defining accessors would help clean this up?  Like

```
def clientkey
  options && options[:clientkey]
end

def clientkey=(val)
  options ||= {}
  options[:clientkey] = val
end
```

Then we probably don't need to override `params_to_attributes` at all.  This would be a lot of change maybe something for a followup

Reply:
yup, followup is good